### PR TITLE
TELCODOCS-1848: Add PreflightValidation v1beta2

### PR DESCRIPTION
**Pull request by StephenJamesSmith**
D/S Docs:[MGMT-17268] Add PreflightValidation v1beta2

Jira: https://issues.redhat.com/browse/TELCODOCS-1848?src=confmacro

Version(s): openshift-4.14, openshift-4.15, openshift-4.16

Fix version: KMMO 2.1

Link to docs preview: https://77202--ocpdocs-pr.netlify.app/openshift-enterprise/latest/updating/preparing_for_updates/kmm-preflight-validation.html

Dev review: @qbarrand
QE review: @cdvultur

### CODE REVIEW

**Comment by StephenJamesSmith** [comment]
@qbarrand @cdvultur Please review and ACK. Thx!

**Review by cdvultur — CHANGES_REQUESTED** [review]
The downstream documentation should describe the `PreflightValidationOCP` and not `PreflightValidation`

I would suggest to re-look over all this change and update it with the good references to PreflightValidationOCP.

File: modules/kmm-build-validation-stage.adoc [hunk 1]
```diff
@@ -6,20 +6,11 @@
...
 
-Build validation is executed only when image validation has failed and there is a `build` section in the `Module` that is relevant for the upgraded kernel. Build validation attempts to run the build job and validate that it finishes successfully.
+Build validation is executed only when image validation has failed and there is a `build` section in the `Module` that is relevant for the upgraded kernel. Build validation attempts to run the build job and validate that it finishes successfully. If the `PushBuiltImage` flag is defined in the `PreflightValidation` custom resource (CR), it also tries to push the resulting image into its repository. The resulting image name is taken from the definition of the `containerImage` field of the `Module` CR.
 
 [NOTE]
 ====
-You must specify the kernel version when running `depmod`, as shown here:
-[source,terminal]
-----
-$ RUN depmod -b /opt ${KERNEL_VERSION}
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-build-validation-stage.adoc:16]: I used this link for updates. (https://deploy-preview-794--kubernetes-sigs-kmm.netlify.app/documentation/preflight_validation/). Of course, this is midstream, so PreflightValidationOCP may still be the valid field for downstream. @qbarrand - please confirm. Also, please confirm putting the `depmod` command back in.

[the remark this answers is not in the document]
Comment (qbarrand) [reply modules/kmm-build-validation-stage.adoc:16]: @StephenJamesSmith Costi is spot on - we should use `PreflightValidationOCP` downstream. `depmod` also belongs here.

File: modules/kmm-example-cr.adoc [hunk 2]
```diff
@@ -4,26 +4,22 @@
...
-
-[source,terminal]
-----
-quay.io/openshift-release-dev/ocp-release@sha256:22e149142517dfccb47be828f012659b1ccf71d26620e6f62468c264a7ce7863
-----
-
-Because `.spec.pushBuiltImage` is set to `true`, KMM pushes the resulting images of Build/Sign into the defined repositories.
+The example verifies all the currently present modules against the upcoming
+`5.8.18-101.fc31.x86_64` kernel, and push the resulting images of
+Build/Sign into the defined repositories.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:14]: fixed.

File: modules/kmm-sign-validation-stage.adoc [hunk 3]
```diff
@@ -6,13 +6,11 @@
 [id="kmm-sign-validation-stage_{context}"]
 = Sign validation stage
 
-Sign validation is executed only when image validation has failed, there is a `sign` section in the `Module` that is relevant for the upgrade kernel, and build validation finished successfully in the event there was a `build` section in the `Module` relevant for the upgraded kernel. Sign validation will try to run the sign job and validate that it finishes successfully.
+Sign validation is executed only when image validation has failed. There is a `sign` section in the `Module` resource that is relevant for the upgrade kernel, and build validation finishes successfully in the event there was a `build` section in the `Module` relevant for the upgraded kernel. Sign validation attempts to run the sign job and validate that it finishes successfully.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-sign-validation-stage.adoc:9]: fixed.

File: modules/kmm-build-validation-stage.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-build-validation-stage.adoc:16]: @qbarrand @cdvultur Please check for changes to all instances of `PreflightValidationOCP` and addition of `depmod` command and ACK.

File: modules/kmm-example-cr.adoc [hunk 4]
```diff
@@ -8,22 +8,15 @@
...
-apiVersion: kmm.sigs.x-k8s.io/v1beta1
+apiVersion: kmm.sigs.x-k8s.io/v1beta2
 kind: PreflightValidationOCP
 metadata:
- name: preflight
+  name: preflight
 spec:
- releaseImage: quay.io/openshift-release-dev/ocp-release@sha256:22e149142517dfccb47be828f012659b1ccf71d26620e6f62468c264a7ce7863
- pushBuiltImage: true
+  kernelVersion: 5.8.18-101.fc31.x86_64
```
Comment (qbarrand) [inline modules/kmm-example-cr.adoc:20]: `PreflightValidationOCP` uses `releaseImage` in its spec, not `kernelVersion`. I would also stick to the existing content here - although increasing the indentation by 1 space is correct.

File: modules/kmm-example-cr.adoc [hunk 5]
```diff
@@ -8,22 +8,15 @@
...
 
-The example verifies all the currently present modules against the upcoming kernel version included in the {product-title} release 4.11.18, which the following release image points to:
-
-[source,terminal]
-----
-quay.io/openshift-release-dev/ocp-release@sha256:22e149142517dfccb47be828f012659b1ccf71d26620e6f62468c264a7ce7863
-----
-
-Because `.spec.pushBuiltImage` is set to `true`, KMM pushes the resulting images of Build/Sign into the defined repositories.
+The example verifies all the currently present modules against the upcoming `5.8.18-101.fc31.x86_64` kernel, and push the resulting images of Build/Sign in to the defined repositories.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:11]: Please verify this change.

File: modules/kmm-build-validation-stage.adoc
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-build-validation-stage.adoc:1]: I had to wrap some lines in this module.

File: modules/kmm-example-cr.adoc (hunk 4 above)
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:20]: Please verify this change.

File: modules/kmm-validation-kickoff.adoc [hunk 6]
```diff
@@ -6,21 +6,10 @@
...
-	// releaseImage describes the OCP release image that all Modules need to be checked against.
-	// +kubebuilder:validation:Required
-	ReleaseImage string `json:"releaseImage"` <1>
-	// Boolean flag that determines whether images build during preflight must also
-	// be pushed to a defined repository
-	// +optional
-	PushBuiltImage bool `json:"pushBuiltImage"` <2>
-}
-----
+`kernelVersion`:: The version of the kernel that the cluster is upgraded to. This field is required.
```
Comment (cdvultur) [inline modules/kmm-validation-kickoff.adoc:12]: This section also requires revert. In Downstream we specify the `releaseImage` , not the `kernelVersion`

**Review by cdvultur — CHANGES_REQUESTED** [review]
Some small changes needed, to reflect Downstream implementation

File: modules/kmm-example-cr.adoc [hunk 7]
```diff
@@ -8,22 +8,15 @@
 
 This section shows an example of the `PreflightValidationOCP` resource in the YAML format.
 
-The example verifies all the currently present modules against the upcoming kernel version included in the {product-title} release 4.11.18, which the following release image points to:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:11]: Added.

File: modules/kmm-validation-kickoff.adoc (hunk 6 above)
Comment (StephenJamesSmith) [reply modules/kmm-validation-kickoff.adoc]: @qbarrand Done. Please check and /lgtm if all changes are complete.

File: modules/kmm-validation-status.adoc [hunk 8]
```diff
@@ -6,43 +6,23 @@
...
-The following fields apply to each module:
-
-<1> `VerificationStatus` - `true` or `false`, validated or not.
-
-<2> `StatusReason` - Verbal explanation regarding the status.
-
-<3> `VerificationStage` - Describes the validation stage being executed (Image, Build, Sign).
-
-<4> `LastTransitionTime` - The time of the last update to the status.
+A `PreflightValidation` resource reports the status and progress of each module in the cluster that it attempts or has attempted to validate in its `.status.modules` list. Elements of that list contain the following fields:
```
Comment (cdvultur) [inline modules/kmm-validation-status.adoc:9]: Maybe worth updating to `PreflightValidationOCP` , to be consistent with the rest of the document.

**Review by cdvultur — CHANGES_REQUESTED** [review]
Small comment, so that all references points to PreflightValidationOCP

File: modules/kmm-validation-status.adoc (hunk 8 above)
Comment (StephenJamesSmith) [reply modules/kmm-validation-status.adoc:9]: Dang! I missed that one.

**Review by cdvultur — COMMENTED** [review]
/lgtm

File: modules/kmm-example-cr.adoc [hunk 9]
```diff
@@ -8,22 +8,22 @@
...
+The example verifies all the currently present modules against the upcoming kernel version included in the OpenShift Container Platform release 4.11.18, which the following release image points to:
```
Comment (stevsmit) [inline modules/kmm-example-cr.adoc:11]:
```suggestion
The example verifies all of the currently present modules against the upcoming kernel version included in the {product-title} release 4.11.18, which the following release image points to:
```

or maybe

```suggestion
The example verifies all currently present modules against the upcoming kernel version included in the {product-title}  release 4.11.18, which the following release image points to:
```

File: modules/kmm-sign-validation-stage.adoc [hunk 10]
```diff
@@ -6,11 +6,9 @@
...
+Sign validation is executed only when image validation has failed. There is a `sign` section in the `Module` resource that is relevant for the upgrade kernel, and build validation finishes successfully in case there was a `build` section in the `Module` relevant for the upgraded kernel. Sign validation attempts to run the sign job and validate that it finishes successfully.
 
-If the `PushBuiltImage` flag is defined in the `PreflightValidationOCP` CR, sign validation will also try to push the resulting image to its registry.
-
-The resulting image is always the image defined in the `containerImage` field of the `Module`. The input image is either the output of the Build stage, or an image defined in the `UnsignedImage` field.
+If the `PushBuiltImage` flag is defined in the `PreflightValidationOCP` CR, sign validation will also try to push the resulting image to its registry. The resulting image is always the image defined in the `ContainerImage` field of the `Module`. The input image is either the output of the Build stage, or an image defined in the `UnsignedImage` field.
```
Comment (stevsmit) [inline modules/kmm-sign-validation-stage.adoc:11]: You've removed a few cases of "will" in this PR. Do you want to remove it here as well? 

```suggestion
If the `PushBuiltImage` flag is defined in the `PreflightValidationOCP` CR, sign validation also tries to push the resulting image to its registry. The resulting image is always the image defined in the `ContainerImage` field of the `Module`. The input image is either the output of the Build stage, or an image defined in the `UnsignedImage` field.
```

File: modules/kmm-validation-kickoff.adoc [hunk 11]
```diff
@@ -8,19 +8,6 @@
...
-	// releaseImage describes the OCP release image that all Modules need to be checked against.
-	// +kubebuilder:validation:Required
-	ReleaseImage string `json:"releaseImage"` <1>
-	// Boolean flag that determines whether images build during preflight must also
-	// be pushed to a defined repository
-	// +optional
-	PushBuiltImage bool `json:"pushBuiltImage"` <2>
-}
-----
+`releaseImage`:: Mandatory field that provides the name of the release image for the {product-title} version the cluster is upgraded to.
```
Comment (stevsmit) [inline modules/kmm-validation-kickoff.adoc:11]:
```suggestion
`releaseImage`:: Mandatory field that provides the name of the release image for the {product-title} version that the cluster is upgraded to.
```

File: modules/kmm-validation-lifecycle.adoc [hunk 12]
```diff
@@ -6,6 +6,6 @@
 [id="kmm-validation-lifecycle_{context}"]
 = Validation lifecycle
 
-Preflight validation attempts to validate every module loaded in the cluster. Preflight will stop running validation on a `Module` resource after the validation is successful. In case module validation has failed, you can change the module definitions and Preflight will try to validate the module again in the next loop.
+Preflight validation attempts to validate every module loaded in the cluster. Preflight stops running validation on a `Module` resource after the validation is successful. If case module validation fails, you can change the module definitions and Preflight will try to validate the module again in the next loop.
```
Comment (stevsmit) [inline modules/kmm-validation-lifecycle.adoc:9]: Same thought here: 

```suggestion
Preflight validation attempts to validate every module loaded in the cluster. Preflight stops running validation on a `Module` resource after the validation is successful. If case module validation fails, you can change the module definitions and Preflight tries to validate the module again in the next loop.
```

**Comment by stevsmit** [comment]
@StephenJamesSmith mostly smaller stuff. Good work!

File: modules/kmm-example-cr.adoc (hunk 9 above)
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:11]: Used the first change.

File: modules/kmm-sign-validation-stage.adoc (hunk 10 above)
Comment (StephenJamesSmith) [reply modules/kmm-sign-validation-stage.adoc:11]: changed.

File: modules/kmm-validation-kickoff.adoc (hunk 11 above)
Comment (StephenJamesSmith) [reply modules/kmm-validation-kickoff.adoc:11]: changed.

File: modules/kmm-validation-lifecycle.adoc (hunk 12 above)
Comment (StephenJamesSmith) [reply modules/kmm-validation-lifecycle.adoc:9]: Changed. s/If case module validation fails/If module validation fails

**Pull request merged**